ICE: bound types encountered in super_relate_tys

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(non_lifetime_binders)]
#![crate_type = "lib"]

pub trait Bt {
    type C;
}

pub struct A;

impl Bt for A {
    type C = A;
}

pub fn oopsie<Bt>()
where
    for<'a, B> <B::C as Bt<'a>>::Bt: Bt, {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (7b4f48927 2023-03-12)
binary: rustc
commit-hash: 7b4f48927dce585f747a58083b45ab62b9d73a53
commit-date: 2023-03-12
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

### Error output

```
error[E0576]: cannot find associated type `Bt` in `Bt`
  --> 7CE45A32AFF0AC746EABE2C16AC685FC843FA4D22A2D9B4BD55181B090556132.rs:16:34
   |
5  |     type C;
   |     ------- associated type `C` defined here
...
16 |     for<'a, B> <B::C as Bt<'a>>::Bt: Bt, {}
   |                                  ^^
   |                                  |
   |                                  not found in `Bt`
   |                                  help: maybe you meant this associated type: `C`

error[E0404]: expected trait, found type parameter `Bt`
  --> 7CE45A32AFF0AC746EABE2C16AC685FC843FA4D22A2D9B4BD55181B090556132.rs:16:38
   |
4  | pub trait Bt {
   |           -- you might have meant to refer to this trait
...
14 | pub fn oopsie<Bt>()
   |               -- found this type parameter
15 | where
16 |     for<'a, B> <B::C as Bt<'a>>::Bt: Bt, {}
   |                                      ^^ not a trait

warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
 --> 7CE45A32AFF0AC746EABE2C16AC685FC843FA4D22A2D9B4BD55181B090556132.rs:1:12
  |
1 | #![feature(non_lifetime_binders)]
  |            ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
  = note: `#[warn(incomplete_features)]` on by default
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```

error: internal compiler error: /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/compiler/rustc_middle/src/ty/relate.rs:413:13: bound types encountered in super_relate_tys

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/compiler/rustc_errors/src/lib.rs:1644:9
stack backtrace:
   0:     0x7f784176651a - std::backtrace_rs::backtrace::libunwind::trace::hf7d75cca16fd1b61
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f784176651a - std::backtrace_rs::backtrace::trace_unsynchronized::hf5ec95734aab509d
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f784176651a - std::sys_common::backtrace::_print_fmt::he95113413355c8cd
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f784176651a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h072d12f2e69ca92b
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f78417c9c5e - core::fmt::write::h1d23f16415e96af2
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/core/src/fmt/mod.rs:1232:17
   5:     0x7f7841759275 - std::io::Write::write_fmt::hcf1e413fde4ffcd5
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/io/mod.rs:1684:15
   6:     0x7f78417662e5 - std::sys_common::backtrace::_print::h6b2b14bedd5177e7
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f78417662e5 - std::sys_common::backtrace::print::h09ccebfbcd382e89
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f784176905f - std::panicking::default_hook::{{closure}}::h4ee055db778febc2
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/panicking.rs:271:22
   9:     0x7f7841768d9b - std::panicking::default_hook::ha74f60e3a15d95fb
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/panicking.rs:290:9
  10:     0x7f7844a97085 - rustc_driver_impl[d355a67661263e47]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f784176989d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h7f34603beb945e8f
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/alloc/src/boxed.rs:2002:9
  12:     0x7f784176989d - std::panicking::rust_panic_with_hook::h4b5d379936d4932f
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/panicking.rs:696:13
  13:     0x7f7845002371 - std[e52d7c968ed01318]::panicking::begin_panic::<rustc_errors[45159b7f7e36021d]::ExplicitBug>::{closure#0}
  14:     0x7f7844ffdaa6 - std[e52d7c968ed01318]::sys_common::backtrace::__rust_end_short_backtrace::<std[e52d7c968ed01318]::panicking::begin_panic<rustc_errors[45159b7f7e36021d]::ExplicitBug>::{closure#0}, !>
  15:     0x7f7845086046 - std[e52d7c968ed01318]::panicking::begin_panic::<rustc_errors[45159b7f7e36021d]::ExplicitBug>
  16:     0x7f7845051e06 - std[e52d7c968ed01318]::panic::panic_any::<rustc_errors[45159b7f7e36021d]::ExplicitBug>
  17:     0x7f784504d966 - <rustc_errors[45159b7f7e36021d]::HandlerInner>::bug::<&alloc[fc860cae0dfa1ab1]::string::String>
  18:     0x7f784504d630 - <rustc_errors[45159b7f7e36021d]::Handler>::bug::<&alloc[fc860cae0dfa1ab1]::string::String>
  19:     0x7f7845039a9b - rustc_middle[a96ac5d591578447]::util::bug::opt_span_bug_fmt::<rustc_span[f8bdfaef83e46314]::span_encoding::Span>::{closure#0}
  20:     0x7f78450384fa - rustc_middle[a96ac5d591578447]::ty::context::tls::with_opt::<rustc_middle[a96ac5d591578447]::util::bug::opt_span_bug_fmt<rustc_span[f8bdfaef83e46314]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f78450384c6 - rustc_middle[a96ac5d591578447]::ty::context::tls::with_context_opt::<rustc_middle[a96ac5d591578447]::ty::context::tls::with_opt<rustc_middle[a96ac5d591578447]::util::bug::opt_span_bug_fmt<rustc_span[f8bdfaef83e46314]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f78450399e6 - rustc_middle[a96ac5d591578447]::util::bug::opt_span_bug_fmt::<rustc_span[f8bdfaef83e46314]::span_encoding::Span>
  23:     0x7f78430b44b3 - rustc_middle[a96ac5d591578447]::util::bug::bug_fmt
  24:     0x7f7842af5f4c - rustc_middle[a96ac5d591578447]::ty::relate::super_relate_tys::<rustc_infer[8071ebb96624e335]::infer::equate::Equate>
  25:     0x7f7842af22b0 - <rustc_infer[8071ebb96624e335]::infer::equate::Equate as rustc_middle[a96ac5d591578447]::ty::relate::TypeRelation>::tys
  26:     0x7f78439cfcf5 - <rustc_infer[8071ebb96624e335]::infer::at::At>::eq::<rustc_middle[a96ac5d591578447]::ty::Ty>
  27:     0x7f78439d0343 - <rustc_infer[8071ebb96624e335]::infer::InferCtxt>::can_eq::<rustc_middle[a96ac5d591578447]::ty::Ty>
  28:     0x7f7844bc5e4f - <&mut core[125d846441ac421f]::iter::adapters::cloned::clone_try_fold<rustc_span[f8bdfaef83e46314]::def_id::DefId, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<()>, core[125d846441ac421f]::iter::traits::iterator::Iterator::any::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}::{closure#1}>::{closure#0}>::{closure#0} as core[125d846441ac421f]::ops::function::FnMut<((), &rustc_span[f8bdfaef83e46314]::def_id::DefId)>>::call_mut
  29:     0x7f7844c0d55f - <core[125d846441ac421f]::iter::adapters::map::Map<indexmap[7ada2b4ea418fd7]::map::Iter<rustc_middle[a96ac5d591578447]::ty::fast_reject::SimplifiedType, alloc[fc860cae0dfa1ab1]::vec::Vec<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, <rustc_middle[a96ac5d591578447]::ty::context::TyCtxt>::all_impls::{closure#0}> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::<(), <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _>>::iter_try_fold::flatten<&alloc[fc860cae0dfa1ab1]::vec::Vec<rustc_span[f8bdfaef83e46314]::def_id::DefId>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<()>, <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::flatten<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<()>, core[125d846441ac421f]::iter::adapters::cloned::clone_try_fold<rustc_span[f8bdfaef83e46314]::def_id::DefId, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<()>, core[125d846441ac421f]::iter::traits::iterator::Iterator::any::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}::{closure#1}>::{closure#0}>::{closure#0}>::{closure#0}>::{closure#0}, core[125d846441ac421f]::ops::control_flow::ControlFlow<()>>
  30:     0x7f7844c2054d - <core[125d846441ac421f]::iter::adapters::chain::Chain<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>, core[125d846441ac421f]::iter::adapters::flatten::FlatMap<indexmap[7ada2b4ea418fd7]::map::Iter<rustc_middle[a96ac5d591578447]::ty::fast_reject::SimplifiedType, alloc[fc860cae0dfa1ab1]::vec::Vec<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, &alloc[fc860cae0dfa1ab1]::vec::Vec<rustc_span[f8bdfaef83e46314]::def_id::DefId>, <rustc_middle[a96ac5d591578447]::ty::context::TyCtxt>::all_impls::{closure#0}>> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::<(), core[125d846441ac421f]::iter::adapters::cloned::clone_try_fold<rustc_span[f8bdfaef83e46314]::def_id::DefId, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<()>, core[125d846441ac421f]::iter::traits::iterator::Iterator::any::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}::{closure#1}>::{closure#0}>::{closure#0}, core[125d846441ac421f]::ops::control_flow::ControlFlow<()>>
  31:     0x7f7844bbfcec - <&mut <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0} as core[125d846441ac421f]::ops::function::FnMut<(&rustc_span[f8bdfaef83e46314]::def_id::DefId,)>>::call_mut
  32:     0x7f7844bab2ea - <&mut core[125d846441ac421f]::iter::traits::iterator::Iterator::find::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, &mut <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}>::{closure#0} as core[125d846441ac421f]::ops::function::FnMut<((), rustc_span[f8bdfaef83e46314]::def_id::DefId)>>::call_mut
  33:     0x7f7844bbc9b3 - <core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::<(), &mut core[125d846441ac421f]::iter::traits::iterator::Iterator::find::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, &mut <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}>::{closure#0}, core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>>
  34:     0x7f7844c20329 - <core[125d846441ac421f]::iter::adapters::chain::Chain<core[125d846441ac421f]::iter::sources::once::Once<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>, core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>>> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::<(), core[125d846441ac421f]::iter::adapters::map::map_try_fold<rustc_span[f8bdfaef83e46314]::def_id::CrateNum, core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>, <rustc_middle[a96ac5d591578447]::ty::context::TyCtxt>::all_traits::{closure#0}, <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _>>::iter_try_fold::flatten<core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>, <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::flatten<core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>, core[125d846441ac421f]::iter::traits::iterator::Iterator::find::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, &mut <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}>::{closure#0}>::{closure#0}>::{closure#0}>::{closure#0}, core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>>
  35:     0x7f7844c0ce5e - <core[125d846441ac421f]::iter::adapters::map::Map<core[125d846441ac421f]::iter::adapters::chain::Chain<core[125d846441ac421f]::iter::sources::once::Once<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>, core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>>>, <rustc_middle[a96ac5d591578447]::ty::context::TyCtxt>::all_traits::{closure#0}> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::<(), <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _>>::iter_try_fold::flatten<core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>, <core[125d846441ac421f]::iter::adapters::flatten::FlattenCompat<_, _> as core[125d846441ac421f]::iter::traits::iterator::Iterator>::try_fold::flatten<core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, (), core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>, core[125d846441ac421f]::iter::traits::iterator::Iterator::find::check<rustc_span[f8bdfaef83e46314]::def_id::DefId, &mut <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}>::{closure#0}>::{closure#0}>::{closure#0}, core[125d846441ac421f]::ops::control_flow::ControlFlow<rustc_span[f8bdfaef83e46314]::def_id::DefId>>
  36:     0x7f7844bfff50 - <alloc[fc860cae0dfa1ab1]::vec::Vec<alloc[fc860cae0dfa1ab1]::string::String> as alloc[fc860cae0dfa1ab1]::vec::spec_from_iter::SpecFromIter<alloc[fc860cae0dfa1ab1]::string::String, core[125d846441ac421f]::iter::adapters::map::Map<core[125d846441ac421f]::iter::adapters::filter::Filter<core[125d846441ac421f]::iter::adapters::flatten::FlatMap<core[125d846441ac421f]::iter::adapters::chain::Chain<core[125d846441ac421f]::iter::sources::once::Once<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>, core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::CrateNum>>>, core[125d846441ac421f]::iter::adapters::copied::Copied<core[125d846441ac421f]::slice::iter::Iter<rustc_span[f8bdfaef83e46314]::def_id::DefId>>, <rustc_middle[a96ac5d591578447]::ty::context::TyCtxt>::all_traits::{closure#0}>, <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#0}>, <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty::{closure#1}>>>::from_iter
  37:     0x7f7844bc1975 - <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::probe_traits_that_match_assoc_ty
  38:     0x7f784390e8d4 - <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::associated_path_to_ty::{closure#0}
  39:     0x7f78438f75a6 - <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::ast_ty_to_ty_inner::{closure#0}
  40:     0x7f78438f7970 - <dyn rustc_hir_analysis[357075c2c3fc2a7b]::astconv::AstConv>::ast_ty_to_ty_inner::{closure#0}
  41:     0x7f784303d3d5 - rustc_hir_analysis[357075c2c3fc2a7b]::collect::predicates_of::gather_explicit_predicates_of
  42:     0x7f784303b1d8 - rustc_hir_analysis[357075c2c3fc2a7b]::collect::predicates_of::explicit_predicates_of
  43:     0x7f7843036e54 - rustc_query_system[704b80a7bc44cd49]::query::plumbing::try_execute_query::<rustc_query_impl[217bee59c5548bf5]::queries::explicit_predicates_of, rustc_query_impl[217bee59c5548bf5]::plumbing::QueryCtxt>
  44:     0x7f784304df73 - rustc_hir_analysis[357075c2c3fc2a7b]::collect::predicates_defined_on
  45:     0x7f784304c596 - rustc_query_system[704b80a7bc44cd49]::query::plumbing::try_execute_query::<rustc_query_impl[217bee59c5548bf5]::queries::predicates_defined_on, rustc_query_impl[217bee59c5548bf5]::plumbing::QueryCtxt>
  46:     0x7f78436358b5 - rustc_query_system[704b80a7bc44cd49]::query::plumbing::try_execute_query::<rustc_query_impl[217bee59c5548bf5]::queries::predicates_of, rustc_query_impl[217bee59c5548bf5]::plumbing::QueryCtxt>
  47:     0x7f784363455f - <rustc_query_impl[217bee59c5548bf5]::Queries as rustc_middle[a96ac5d591578447]::ty::query::QueryEngine>::predicates_of
  48:     0x7f784327be02 - <rustc_hir_analysis[357075c2c3fc2a7b]::collect::CollectItemTypesVisitor as rustc_hir[4aa57babd8685147]::intravisit::Visitor>::visit_item
  49:     0x7f784327a925 - <rustc_middle[a96ac5d591578447]::hir::map::Map>::visit_item_likes_in_module::<rustc_hir_analysis[357075c2c3fc2a7b]::collect::CollectItemTypesVisitor>
  50:     0x7f784327a84c - rustc_hir_analysis[357075c2c3fc2a7b]::collect::collect_mod_item_types
  51:     0x7f78440c706e - rustc_query_system[704b80a7bc44cd49]::query::plumbing::try_execute_query::<rustc_query_impl[217bee59c5548bf5]::queries::collect_mod_item_types, rustc_query_impl[217bee59c5548bf5]::plumbing::QueryCtxt>
  52:     0x7f78440c6bf3 - <rustc_query_impl[217bee59c5548bf5]::Queries as rustc_middle[a96ac5d591578447]::ty::query::QueryEngine>::collect_mod_item_types
  53:     0x7f7842db0363 - <rustc_session[7e23441c2db8ce4a]::session::Session>::track_errors::<rustc_hir_analysis[357075c2c3fc2a7b]::check_crate::{closure#0}, ()>
  54:     0x7f7842dacce2 - rustc_hir_analysis[357075c2c3fc2a7b]::check_crate
  55:     0x7f7842da2d12 - rustc_interface[6bf42eb308c32713]::passes::analysis
  56:     0x7f78442ab71c - rustc_query_system[704b80a7bc44cd49]::query::plumbing::try_execute_query::<rustc_query_impl[217bee59c5548bf5]::queries::analysis, rustc_query_impl[217bee59c5548bf5]::plumbing::QueryCtxt>
  57:     0x7f78442ab410 - <rustc_query_impl[217bee59c5548bf5]::Queries as rustc_middle[a96ac5d591578447]::ty::query::QueryEngine>::analysis
  58:     0x7f78440d02f9 - <rustc_middle[a96ac5d591578447]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[d355a67661263e47]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>>
  59:     0x7f7843cc8498 - rustc_span[f8bdfaef83e46314]::with_source_map::<core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>, rustc_interface[6bf42eb308c32713]::interface::run_compiler<core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>, rustc_driver_impl[d355a67661263e47]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  60:     0x7f7843cbf70c - std[e52d7c968ed01318]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[6bf42eb308c32713]::util::run_in_thread_pool_with_globals<rustc_interface[6bf42eb308c32713]::interface::run_compiler<core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>, rustc_driver_impl[d355a67661263e47]::run_compiler::{closure#1}>::{closure#0}, core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>>
  61:     0x7f7843cbf13a - <<std[e52d7c968ed01318]::thread::Builder>::spawn_unchecked_<rustc_interface[6bf42eb308c32713]::util::run_in_thread_pool_with_globals<rustc_interface[6bf42eb308c32713]::interface::run_compiler<core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>, rustc_driver_impl[d355a67661263e47]::run_compiler::{closure#1}>::{closure#0}, core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[125d846441ac421f]::result::Result<(), rustc_span[f8bdfaef83e46314]::ErrorGuaranteed>>::{closure#1} as core[125d846441ac421f]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  62:     0x7f7841773793 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc0a50816e1eb2bba
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/alloc/src/boxed.rs:1988:9
  63:     0x7f7841773793 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb2f83fc2049033c3
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/alloc/src/boxed.rs:1988:9
  64:     0x7f7841773793 - std::sys::unix::thread::Thread::new::thread_start::hce5861512f8bf7ea
                               at /rustc/7b4f48927dce585f747a58083b45ab62b9d73a53/library/std/src/sys/unix/thread.rs:108:17
  65:     0x7f7841505bb5 - <unknown>
  66:     0x7f7841587d90 - <unknown>
  67:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.70.0-nightly (7b4f48927 2023-03-12) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [explicit_predicates_of] computing explicit predicates of `oopsie`
#1 [predicates_defined_on] computing predicates of `oopsie`
#2 [predicates_of] computing predicates of `oopsie`
#3 [collect_mod_item_types] collecting item types in top-level module
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors; 1 warning emitted

Some errors have detailed explanations: E0404, E0576.
For more information about an error, try `rustc --explain E0404`.
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/astconv/mod.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/mod.rs
@@ -2396,13 +2396,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                                     tcx,
                                     infcx.fresh_substs_for_item(DUMMY_SP, impl_def_id),
                                 );
+                                // I guess we don't need to make a universe unless we need it,
+                                // but also we're on the error path, so it doesn't matter here.
+                                let universe = infcx.create_next_universe();
                                 infcx
                                     .can_eq(
                                         ty::ParamEnv::empty(),
                                         impl_.self_ty(),
-                                        // Must fold past escaping bound vars too,
-                                        // since we have those at this point in astconv.
-                                        tcx.fold_regions(qself_ty, |_, _| tcx.lifetimes.re_erased),
+                                        tcx.replace_escaping_bound_vars_uncached(qself_ty, ty::fold::FnMutDelegate {
+                                            regions: &mut |_| tcx.lifetimes.re_erased,
+                                            types: &mut |bv| tcx.mk_placeholder(ty::PlaceholderType {
+                                                universe,
+                                                name: bv.kind,
+                                            }),
+                                            consts: &mut |bv, ty| tcx.mk_const(ty::PlaceholderConst {
+                                                universe,
+                                                name: bv
+                                            }, ty),
+                                        })
                                     )
                             })
                             && tcx.impl_polarity(impl_def_id) != ty::ImplPolarity::Negative

--- a/tests/ui/traits/non_lifetime_binders/missing-assoc-item.rs
+++ b/tests/ui/traits/non_lifetime_binders/missing-assoc-item.rs
@@ -1,0 +1,11 @@
+#![feature(non_lifetime_binders)]
+//~^ WARN the feature `non_lifetime_binders` is incomplete
+
+fn f()
+where
+    for<B> B::Item: Send,
+    //~^ ERROR ambiguous associated type
+{
+}
+
+fn main() {}

--- a/tests/ui/traits/non_lifetime_binders/missing-assoc-item.stderr
+++ b/tests/ui/traits/non_lifetime_binders/missing-assoc-item.stderr
@@ -1,0 +1,18 @@
+warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/missing-assoc-item.rs:1:12
+   |
+LL | #![feature(non_lifetime_binders)]
+   |            ^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
+   = note: `#[warn(incomplete_features)]` on by default
+
+error[E0223]: ambiguous associated type
+  --> $DIR/missing-assoc-item.rs:6:12
+   |
+LL |     for<B> B::Item: Send,
+   |            ^^^^^^^ help: use the fully-qualified path: `<B as IntoIterator>::Item`
+
+error: aborting due to previous error; 1 warning emitted
+
+For more information about this error, try `rustc --explain E0223`.

